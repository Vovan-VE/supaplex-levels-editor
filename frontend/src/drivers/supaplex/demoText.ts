// -- High nibble: Time-unit repeat count minus 1 for low nibble command key.
//    This is the information from which the total demo duration
//    is calculated, corrected with the (stored) SpeedFix setting.
//    At normalized speed, these time units are 1/35th of a second
//    each. Example: If this nibble shows 9, it means 9+1=10
//    time units for the command key in the lower nibble.
// -- Low nibble:
//    Command key, equivalent with the keyboard game keys:
//      0 = just wait: no key pressed
//      1 = cursor up
//      2 = cursor left
//      3 = cursor down
//      4 = cursor right
//      5 = space + cursor up
//      6 = space + cursor left
//      7 = space + cursor down
//      8 = space + cursor right
//      9 = space only
//      A = (not used)
//      B = (not used)
//      C = (not used)
//      D = (not used)
//      E = (not used)
//      F = (not used! / low nibble of the end-of-demo byte)
// - hex FFh byte, meaning: end of demo.
//
// The maximum demo file length is 1+64008+1=64010 bytes, which is an arbitrary
// amount set by Robin Heydon (which could have been 65514 bytes maximum only).
//
// This maximum total demo length of 64010 bytes has been reduced in the SpeedFix
// versions 5 and up to 48650 bytes because the same space is needed for all 10
// demo levels of 1536 bytes each. (48650+10*1536=64010) See also just below.

const KEY_STR = ["-", "U", "L", "D", "R", "SU", "SL", "SD", "SR", "S"];
const CH_TILES = ".";
const FRAMES_PER_TILE = 8;

export interface ToTextOptions {
  wrapWidth?: number;
  useTilesTime?: boolean;
}

export const demoToText = (
  demo: Uint8Array | null,
  { wrapWidth = Infinity, useTilesTime = false }: ToTextOptions = {},
): string => {
  if (!demo) {
    return "";
  }
  let text = "";
  let lastKey = 0;
  let duration = 0;
  let lineLen = 0;
  function flush() {
    if (!duration) {
      return;
    }
    if (lastKey < KEY_STR.length) {
      let add = KEY_STR[lastKey];

      if (duration > 1) {
        if (useTilesTime) {
          const tiles = Math.floor(duration / FRAMES_PER_TILE);
          const rest = duration % FRAMES_PER_TILE;
          if (tiles > 0) {
            add += tiles + CH_TILES;
          }
          if (rest > 0) {
            add += rest;
          }
        } else {
          add += duration;
        }
      }

      if (lineLen > 0) {
        if (lineLen + 1 + add.length > wrapWidth) {
          text += "\n";
          lineLen = 0;
        } else {
          text += " ";
          lineLen++;
        }
      }

      text += add;
      lineLen += add.length;
    }
    duration = 0;
  }
  for (const hh of demo) {
    const time = ((hh >> 4) & 0xf) + 1;
    const key = hh & 0xf;
    if (key !== lastKey) {
      flush();
      lastKey = key;
    }
    duration += time;
  }
  flush();

  return text;
};

export const demoFromText = (text: string) => {
  const demo: number[] = [];
  function flush(key: number, duration: number) {
    const fulls = Math.floor(duration / 16);
    const rest = duration % 16;
    if (fulls > 0) {
      const hh = 0xf0 | key;
      for (let i = fulls; i-- > 0; ) {
        demo.push(hh);
      }
    }
    if (rest > 0) {
      demo.push(((rest - 1) << 4) | key);
    }
  }

  const L = text.length;
  let at = 0;
  function parseInt() {
    let to = at;
    let ch: string;
    while ((ch = text.charAt(to)) >= "0" && ch <= "9") {
      to++;
    }
    if (to > at) {
      return { int: Number(text.substring(at, to)), next: to };
    }
    return null;
  }
  while (at < L) {
    let ch = text.charAt(at);
    if (/\s/u.test(ch)) {
      at++;
      continue;
    }
    ch = ch.toUpperCase();

    let to = at;
    if (ch === "-") {
      to++;
    } else {
      if (ch === "S") {
        to++;
        ch = text.charAt(to).toUpperCase();
      }
      switch (ch) {
        case "U":
        case "L":
        case "D":
        case "R":
          to++;
      }
    }
    if (to === at) {
      throw new SyntaxError(
        `Unexpected char ${JSON.stringify(text.charAt(at))} at offset ${at}`,
      );
    }
    const keyStr = text.substring(at, to).toUpperCase();
    const keyCode = KEY_STR.indexOf(keyStr);
    if (keyCode < 0) {
      throw new Error(`bad parse state: keyStr=${JSON.stringify(keyStr)}`);
    }
    at = to;

    const wasAt = at;
    let frames = 0;
    const int1 = parseInt();
    if (int1) {
      at = int1.next;
      if (text.charAt(at) === CH_TILES) {
        at++;
        frames = int1.int * FRAMES_PER_TILE;

        const int2 = parseInt();
        if (int2) {
          at = int2.next;
          frames += int2.int;
        }
      } else {
        frames = int1.int;
      }
    } else {
      frames = 1;
    }

    if (!frames) {
      throw new SyntaxError(`Invalid zero duration at offset ${wasAt}`);
    }

    flush(keyCode, frames);
  }

  if (demo.length <= 0) {
    return null;
  }
  return new Uint8Array(demo);
};
