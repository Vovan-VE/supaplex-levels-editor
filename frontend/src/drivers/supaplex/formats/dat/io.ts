import { createLevel } from "../../level";
import { createLevelset } from "../../levelset";
import { ISupaplexLevel, ISupaplexLevelset } from "../../types";
import { LEVEL_BYTES_LENGTH, LEVEL_HEIGHT, LEVEL_WIDTH } from "../std";

function* levelsFromBuffer(buffer: ArrayBuffer): Iterable<ISupaplexLevel> {
  if (buffer.byteLength % LEVEL_BYTES_LENGTH) {
    throw new Error("Invalid file size: not a module of level size");
  }

  const count = Math.floor(buffer.byteLength / LEVEL_BYTES_LENGTH);
  for (let i = 0; i < count; i++) {
    yield createLevel(
      LEVEL_WIDTH,
      LEVEL_HEIGHT,
      new Uint8Array(buffer, i * LEVEL_BYTES_LENGTH, LEVEL_BYTES_LENGTH),
    );
  }
}

export const readLevelset = (file: ArrayBuffer): ISupaplexLevelset =>
  createLevelset(levelsFromBuffer(file));

export const writeLevelset = (levelset: ISupaplexLevelset): ArrayBuffer => {
  const result = new Uint8Array(levelset.levelsCount * LEVEL_BYTES_LENGTH);
  for (let [i, level] of levelset.getLevels().entries()) {
    // errors covered by "support report"
    if (level.width > LEVEL_WIDTH || level.height > LEVEL_HEIGHT) {
      throw new Error(
        `Level ${i + 1} is too large: ${level.width}x${level.height}`,
      );
    }

    // warnings covered by "support report"
    if (level.width < LEVEL_WIDTH || level.height < LEVEL_HEIGHT) {
      level = level.resize({ width: LEVEL_WIDTH, height: LEVEL_HEIGHT });
    }
    level = level.setDemo(null);

    result.set(level.raw, i * LEVEL_BYTES_LENGTH);
  }
  return result.buffer;
};
