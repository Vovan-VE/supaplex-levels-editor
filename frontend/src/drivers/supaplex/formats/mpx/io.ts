import { createLevel } from "../../level";
import { createLevelset } from "../../levelset";
import { ISupaplexLevel, ISupaplexLevelset } from "../../types";
import { FOOTER_BYTE_LENGTH } from "../std";

// MPX format
// 04 : "MPX "
// 04 : 01 00 01 00
//      -ver- -----
//            levels
//            count
// repeat for every level {
// 04 :  w 00  h 00
// 08 : 15 00 00 00
//      -offset+1--
// 0C : 82 06 00 00
//      ---size----
// }
// repeat for every level {
//    : std sp level with respect to w & h
//    : demo
// }
//
// - https://git.artsoft.org/?p=rocksndiamonds.git;a=commit;h=29ea60b3
// - https://github.com/ArtsoftEntertainment/rocksndiamonds/commit/29ea60b3#diff-6854d2255acd6155f61b1558b39cbce581fad371b897b8ab58a90e6a227a6a5eR370

const MPX_SIGN = new Uint8Array("MPX ".split("").map((ch) => ch.charCodeAt(0)));
const MIN_FILE_SIZE = 8 + 12 + 1 + FOOTER_BYTE_LENGTH;

const cmpUint8Array = (
  buffer: ArrayBuffer,
  byteOffset: number,
  match: Uint8Array,
) =>
  new Uint8Array(buffer, byteOffset, match.length).every(
    (b, i) => b === match[i],
  );

const getInt16LE = (file: ArrayBuffer, byteOffset: number) =>
  new DataView(file, byteOffset, 2).getInt16(0, true);

const getInt32LE = (file: ArrayBuffer, byteOffset: number) =>
  new DataView(file, byteOffset, 4).getInt32(0, true);

const setInt16LE = (buffer: Uint8Array, byteOffset: number, value: number) => {
  new DataView(buffer.buffer).setInt16(byteOffset, value, true);
};

const setInt32LE = (buffer: Uint8Array, byteOffset: number, value: number) => {
  new DataView(buffer.buffer).setInt32(byteOffset, value, true);
};

export const readLevelset = (file: ArrayBuffer): ISupaplexLevelset => {
  if (file.byteLength < MIN_FILE_SIZE || !cmpUint8Array(file, 0, MPX_SIGN)) {
    throw new Error("Unrecognized file format");
  }

  const ver = getInt16LE(file, 4);
  let levelsCount = getInt16LE(file, 6);
  const widths: number[] = [];
  const heights: number[] = [];
  const offsets: number[] = [];
  const sizes: number[] = [];
  if (ver === 0x2020 && levelsCount === 0x2020) {
    // 4 spaces
    //ver = 1;
    levelsCount = 1;
    widths.push(getInt16LE(file, 8));
    heights.push(getInt16LE(file, 10));
    offsets.push(20);
    sizes.push(widths[0] * heights[0] + FOOTER_BYTE_LENGTH);
  } else {
    if (ver !== 1) {
      throw new Error(`Unsupported format version ${ver}`);
    }
    for (let i = 0, at = 8; i < levelsCount; i++, at += 12) {
      widths.push(getInt16LE(file, at));
      heights.push(getInt16LE(file, at + 2));
      offsets.push(getInt32LE(file, at + 4) - 1);
      sizes.push(getInt32LE(file, at + 8));
    }
  }

  const levels: ISupaplexLevel[] = [];
  for (let i = 0; i < levelsCount; i++) {
    levels.push(
      createLevel(
        widths[i],
        heights[i],
        new Uint8Array(file, offsets[i], sizes[i]),
      ),
    );
  }

  return createLevelset(levels);
};

export const writeLevelset = (levelset: ISupaplexLevelset): ArrayBuffer => {
  // warnings covered by "support report"
  const levelsCount = Math.min(0x7fff, levelset.levelsCount);
  const levels = levelset.getLevels().slice(0, levelsCount);

  const widths: number[] = [];
  const heights: number[] = [];
  const offsets: number[] = [];
  const sizes: number[] = [];

  let offset = 8 + 12 * levelsCount;
  for (const [i, level] of levels.entries()) {
    // errors covered by "support report"
    if (level.width > 0x7fff || level.height > 0x7fff) {
      throw new Error(
        `Level ${i + 1} is too large: ${level.width}x${level.height}`,
      );
    }
    if (offset >= 0x7fffffff - 1) {
      throw new Error(
        `Level ${
          i + 1
        } cannot be saved due to MPX file format limitations (offset+1 exceeds maximum)`,
      );
    }

    widths.push(level.width);
    heights.push(level.height);
    offsets.push(offset);
    sizes.push(level.length);
    offset += level.length;
  }

  const result = new Uint8Array(offset);
  result.set(MPX_SIGN, 0);
  setInt16LE(result, 4, 1);
  setInt16LE(result, 6, levelsCount);
  for (let i = 0, at = 8; i < levelsCount; i++, at += 12) {
    setInt16LE(result, at, widths[i]);
    setInt16LE(result, at + 2, heights[i]);
    setInt32LE(result, at + 4, offsets[i] + 1);
    setInt32LE(result, at + 8, sizes[i]);

    result.set(levels[i].raw, offsets[i]);
  }

  return result.buffer;
};
