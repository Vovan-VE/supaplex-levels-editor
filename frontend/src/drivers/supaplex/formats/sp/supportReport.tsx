import { Trans } from "i18n/Trans";
import { ISupportReportMessage } from "../../../types";
import { ISupaplexLevel, ISupaplexLevelset } from "../../types";
import { LEVEL_HEIGHT, LEVEL_WIDTH } from "../std";
import { err, warn } from "../supportLevelsetReporter";

export function* supportReport(levelset: ISupaplexLevelset) {
  yield* levelsetReporter(levelset);
  for (const item of levelReporter(levelset.getLevel(0))) {
    yield { ...item, levelIndex: 0 };
  }
}

export function* levelsetReporter(
  levelset: ISupaplexLevelset,
): Iterable<ISupportReportMessage> {
  if (levelset.levelsCount > 1) {
    yield warn(
      <Trans
        i18nKey="main:supaplex.convert.OnlyFirstLevelWillLeft"
        values={{ rest: levelset.levelsCount - 1 }}
      />,
    );
  }
}

export function* levelReporter(
  level: ISupaplexLevel,
): Iterable<ISupportReportMessage> {
  const { width, height } = level;
  if (width > LEVEL_WIDTH || height > LEVEL_HEIGHT) {
    yield err(
      <Trans
        i18nKey="main:supaplex.convert.BodySizeExceed"
        values={{
          width,
          height,
          maxWidth: LEVEL_WIDTH,
          maxHeight: LEVEL_HEIGHT,
        }}
      />,
    );
  } else if (width < LEVEL_WIDTH || height < LEVEL_HEIGHT) {
    yield warn(
      <Trans
        i18nKey="main:supaplex.convert.BodySizeWillExtend"
        values={{
          width,
          height,
          maxWidth: LEVEL_WIDTH,
          maxHeight: LEVEL_HEIGHT,
        }}
      />,
    );
  }
}
