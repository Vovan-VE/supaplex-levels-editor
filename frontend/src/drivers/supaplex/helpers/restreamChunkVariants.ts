import { ITilesStreamItem } from "../../types";

export function* restreamChunkVariants(
  chunk: ITilesStreamItem,
  getVariant: (x: number, y: number) => number | undefined,
): Iterable<ITilesStreamItem> {
  const [x, y, width, tile] = chunk;
  let lastItem: [
    x: number,
    y: number,
    width: number,
    tile: number,
    variant: number | undefined,
  ] = [x, y, 0, tile, undefined];
  for (let i = 0; i < width; i++) {
    const cx = x + i;
    const variant = getVariant(cx, y);
    if (variant === lastItem[4]) {
      lastItem[2]++;
    } else {
      if (lastItem[2]) {
        yield lastItem;
      }
      lastItem = [cx, y, 1, tile, variant];
    }
  }
  if (lastItem[2]) {
    yield lastItem;
  }
}
