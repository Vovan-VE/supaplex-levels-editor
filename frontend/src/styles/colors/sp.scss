$color-black: #000000;
$color-grey-xdark: #616161;
$color-grey-dark: #717171;
$color-grey: #828282;
$color-grey-light: #929292;
$color-grey-xlight: #b2b2b2;
$color-white: #f3f3f3;
$color-red-light: #e31010;
$color-red-dark: #b20000;
$color-brown: #824100;
$color-orange: #f39241;
$color-yellow: #e3e300;
$color-green-dark: #008251;
$color-green-light: #00b261;
$color-blue-dark: #5151d3;
$color-blue-light: #7192e3;
