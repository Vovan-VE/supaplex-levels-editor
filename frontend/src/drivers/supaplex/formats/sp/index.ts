import { ISupaplexFormat } from "../../types";
import { DAT } from "../dat";
import { readLevelset, writeLevelset } from "./io";
import { createLevelset } from "../../levelset";
import { SIGNATURE_MAX_LENGTH } from "../std";
import { supportReport } from "./supportReport";

export const SP: ISupaplexFormat = {
  title: "SP",
  fileExtensionDefault: "sp",
  resizable: DAT.resizable,
  minLevelsCount: 1,
  maxLevelsCount: 1,
  demoSupport: true,
  signatureMaxLength: SIGNATURE_MAX_LENGTH,
  supportReport,
  readLevelset,
  writeLevelset,
  createLevelset: (levels) => createLevelset(levels ?? 1),
  createLevel: DAT.createLevel,
};
