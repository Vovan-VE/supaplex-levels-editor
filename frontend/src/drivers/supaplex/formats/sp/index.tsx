import { Trans } from "i18n/Trans";
import { ISupaplexFormat } from "../../types";
import { DAT } from "../dat";
import { readLevelset, writeLevelset } from "./io";
import { createLevelset } from "../../levelset";
import { SIGNATURE_MAX_LENGTH } from "../std";
import { supportReport } from "./supportReport";

export const SP: ISupaplexFormat = {
  title: "SP",
  description: <Trans i18nKey="main:supaplex.format.sp.desc" />,
  fileExtensionDefault: "sp",
  resizable: DAT.resizable,
  minLevelsCount: 1,
  maxLevelsCount: 1,
  demoSupport: true,
  signatureMaxLength: SIGNATURE_MAX_LENGTH,
  supportReport,
  readLevelset,
  writeLevelset,
  createLevelset: (levels) => createLevelset(levels ?? 1),
  createLevel: DAT.createLevel,
};
