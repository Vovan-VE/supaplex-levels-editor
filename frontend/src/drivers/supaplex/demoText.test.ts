import { demoFromText, demoToText } from "./demoText";
import { readLevelset } from "./formats/sp/io";
import { readExampleFile } from "./helpers.dev";

describe("demo text", () => {
  const getDemo = async (exampleFilename: string) =>
    readLevelset((await readExampleFile(exampleFilename)).buffer).getLevel(0)
      .demo;

  for (const REF of ["D95S001.sp", "D95S018.sp"]) {
    it(`${REF}`, async () => {
      const demo = await getDemo(REF);

      const textDefault = demoToText(demo, {
        wrapWidth: 72,
      });
      const textWithTiles = demoToText(demo, {
        wrapWidth: 72,
        useTilesTime: true,
      });

      expect(textDefault).toMatchSnapshot("default");
      expect(textWithTiles).toMatchSnapshot("+tiles");

      expect(demoFromText(textDefault)).toEqual(demo);
      expect(demoFromText(textWithTiles)).toEqual(demo);
    });
  }
});
