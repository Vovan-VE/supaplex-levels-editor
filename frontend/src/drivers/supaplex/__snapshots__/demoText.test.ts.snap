// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`demo text > D95S001.sp > +tiles 1`] = `
"R2.5 -4 U1.3 - R3.5 - D3.4 R2.5 SR S3 SU4 -2 R5.4 -2 U2. - R1.6 - D7
SD1.1 S -3 U1.5 - R3.6 - U1. - R2.6 SR S3 SD3 S -2 L6.4 - D7 -2 L2.6 U1.
- L5. D1. - L3.4 U1.2 - L5.7 - D3. L3.7 -2.2 R11.6 - D5 - L1.7 - U6 -
L1.7 -2 D5 R7.1 L5.2 -2.3 R7.6 -2 L6 -5 D6 -2 L2.5 - U1. SU1. S3 -1.3 R2
-2.2 L1.7 - U1.5 -1.6 R2 -1.4 S4 SU4 S3 -3.6 R2 -1.6 L2.3 U1.5 -3.4 R
-3.1 S3 SU3 S2 -2 R1. -2 U1.2 -4 L2.6 - D1.1 SD5 S3 SR R6 -2.2 U5 -2
L4.2 -2 D5.1 - R2.1 SR4 S2 -5 L1. U2.6 - L2.1 SL6 S -5 U1.1 -2 L3.6 -
D5. SD2 S SR4 S - L1. - U1.6 -2 L5. SL S1.3 SR4 S2 -3 U3 -2 L2.6 SL S
SU5 S2 -2 L6 - S3 SU5 -2 L1.4 -2 D1.2 SD1. S -6 U2.5 -2 L1.6 -2."
`;

exports[`demo text > D95S001.sp > default 1`] = `
"R21 -4 U11 - R29 - D28 R21 SR S3 SU4 -2 R44 -2 U16 - R14 - D7 SD9 S -3
U13 - R30 - U8 - R22 SR S3 SD3 S -2 L52 - D7 -2 L22 U8 - L40 D8 - L28
U10 - L47 - D24 L31 -18 R94 - D5 - L15 - U6 - L15 -2 D5 R57 L42 -19 R62
-2 L6 -5 D6 -2 L21 - U8 SU8 S3 -11 R2 -18 L15 - U13 -14 R2 -12 S4 SU4 S3
-30 R2 -14 L19 U13 -28 R -25 S3 SU3 S2 -2 R8 -2 U10 -4 L22 - D9 SD5 S3
SR R6 -18 U5 -2 L34 -2 D41 - R17 SR4 S2 -5 L8 U22 - L17 SL6 S -5 U9 -2
L30 - D40 SD2 S SR4 S - L8 - U14 -2 L40 SL S11 SR4 S2 -3 U3 -2 L22 SL S
SU5 S2 -2 L6 - S3 SU5 -2 L12 -2 D10 SD8 S -6 U21 -2 L14 -16"
`;

exports[`demo text > D95S018.sp > +tiles 1`] = `"R7 - D1.7 S2 SL3 SR R1.4 -2. L1.7 -2.4 U2.2 R1.1 - U2.1 - R1.4 -2."`;

exports[`demo text > D95S018.sp > default 1`] = `"R7 - D15 S2 SL3 SR R12 -16 L15 -20 U18 R9 - U17 - R12 -16"`;
